File: 703.kth-largest-element-in-a-stream.cpp
/*
 * @lc app=leetcode id=703 lang=cpp
 *
 * [703] Kth Largest Element in a Stream
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class KthLargest {
 public:
  KthLargest(int k, vector<int>& nums) : k_(k) {
    for (const int num : nums) {
      que_.push(num);
      if (static_cast<int>(que_.size()) > k) {
        que_.pop();
      }
    }
  }

  int add(int val) {
    que_.push(val);
    if (static_cast<int>(que_.size()) > k_) {
      que_.pop();
    }
    return que_.top();
  }

 private:
  const int k_;
  std::priority_queue<int, std::vector<int>, std::greater<int>> que_;
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */
// @lc code=end

int main() {
  vector<int> nums;
  {
    nums = {4, 5, 8, 2};
    KthLargest k(3, nums);
    ASSERT_EQ(k.add(3), 4);
    ASSERT_EQ(k.add(5), 5);
    ASSERT_EQ(k.add(10), 5);
    ASSERT_EQ(k.add(9), 8);
    ASSERT_EQ(k.add(4), 8);
  }
  {
    nums = {};
    KthLargest k(1, nums);
    ASSERT_EQ(k.add(-3), -3);
    ASSERT_EQ(k.add(-2), -2);
    ASSERT_EQ(k.add(-4), -2);
    ASSERT_EQ(k.add(0), 0);
    ASSERT_EQ(k.add(4), 4);
  }
  {
    nums = {0};
    KthLargest k(2, nums);
    ASSERT_EQ(k.add(-1), -1);
    ASSERT_EQ(k.add(1), 0);
    ASSERT_EQ(k.add(-2), 0);
    ASSERT_EQ(k.add(-4), 0);
    ASSERT_EQ(k.add(3), 1);
  }
  return 0;
}

File: 703.kth-largest-element-in-a-stream.rs
/*
 * @lc app=leetcode id=703 lang=rust
 *
 * [703] Kth Largest Element in a Stream
 */

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
struct KthLargest {
    k: i32,
    queue: std::collections::BinaryHeap<std::cmp::Reverse<i32>>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl KthLargest {
    fn new(k: i32, mut nums: Vec<i32>) -> Self {
        nums.sort_unstable();
        nums.reverse();
        let queue = nums
            .iter()
            .take(k as usize)
            .copied()
            .map(std::cmp::Reverse)
            .collect::<std::collections::BinaryHeap<std::cmp::Reverse<i32>>>();
        KthLargest { k, queue }
    }

    fn add(&mut self, val: i32) -> i32 {
        self.queue.push(std::cmp::Reverse(val));
        if self.queue.len() > self.k as usize {
            self.queue.pop();
        }
        match *self.queue.peek().unwrap() {
            std::cmp::Reverse(num) => num,
        }
    }
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * let obj = KthLargest::new(k, nums);
 * let ret_1: i32 = obj.add(val);
 */
// @lc code=end

#[test]
fn test1() {
    let mut obj = KthLargest::new(3, vec![4, 5, 8, 2]);
    assert_eq!(obj.add(3), 4);
    assert_eq!(obj.add(5), 5);
    assert_eq!(obj.add(10), 5);
    assert_eq!(obj.add(9), 8);
    assert_eq!(obj.add(4), 8);
}
